netket/utils/group/semigroup.py

```diff
@@ -0,0 +1,169 @@
+# Copyright 2021 The NetKet Authors - All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Ignore false-positives for redefined `product` functions:
+# pylint: disable=function-redefined
+
+
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+import itertools
+from typing import Callable, List
+
+from plum import dispatch
+
+import numpy as np
+
+from netket.utils import HashableArray
+from netket.utils.types import Array, DType, Shape
+
+
+class ElementBase(ABC):
+    @abstractmethod
+    def __call__(self, arg):
+        pass
+
+    def __matmul__(self, other):
+        return product(self, other)
+
+
+class Element(ElementBase):
+    pass
+
+
+@dataclass(frozen=True)
+class Identity(ElementBase):
+    def __call__(self, arg):
+        return arg
+
+    def __repr__(self):
+        return "Id()"
+
+
+@dispatch
+def product(a: Identity, _: Identity):
+    return a
+
+
+@dispatch
+def product(_: Identity, b: Element):
+    return b
+
+
+@dispatch
+def product(a: Element, _: Identity):
+    return a
+
+
+@dataclass(frozen=True)
+class Composite(Element):
+    left: Element
+    right: Element
+
+    def __call__(self, arg):
+        return self.left(self.right(arg))
+
+    def __repr__(self):
+        return f"{self.left} @ {self.right}"
+
+
+@dispatch
+def product(a: Element, b: Element):
+    return Composite(a, b)
+
+
+@dispatch
+def product(ab: Composite, c: Element):
+    bc = product(ab.right, c)
+    if isinstance(bc, Composite):
+        return Composite(ab, c)
+    else:
+        return Composite(ab.left, bc)
+
+
+@dispatch
+def product(a: Element, bc: Composite):
+    ab = product(a, bc.left)
+    if isinstance(ab, Composite):
+        return Composite(a, bc)
+    else:
+        return Composite(ab, bc.right)
+
+
+@dispatch
+def product(ab: Composite, cd: Composite):
+    bc = product(ab.right, cd.left)
+    if isinstance(bc, Composite):
+        return Composite(ab, cd)
+    else:
+        return Composite(ab.left, Composite(bc, cd.right))
+
+
+@dataclass(frozen=True)
+class NamedElement(Element):
```

Comment:
this is not used anywhere so unless. @femtobit has something against it I'd get rid of it.

Reply:
I have some of my own code that used it because I implemented a custom symmetry and didn't feel like creating an extra class for the element. OTOH, that could be trivially changed so I wouldn't oppose removing it.

Reply:
I'm in favour of leaving `extension points` in.
I feel this feels awkward because what if you want a `NamedPermutationElement`?
is it a named element? a permutation element?
It feels weird to have this in the type hierarchy.
being named is more of a trait.

Reply:
I'd remove this. One can get a named permutation object by, e.g.,
```python3
class CustomPermutation(Permutation):
    def __repr__(self):
        return "CustomPermutation()"
```
which is only marginally more code than creating `NamedElement("CustomPermutation", lambda x: permute(x))` or similar.

Reply:
good.

Reply:
BTW, I changed `Permutation` to allow custom names, since that's how we usually want to see it.